src/wp-includes/general-template.php

```diff
@@ -3413,6 +3413,119 @@ function wp_resource_hints() {
+ *
+ * @link https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/preload
+ * @link https://web.dev/preload-responsive-images/
+ *
+ * @since 6.0.0
+ */
+function wp_preload_links() {
+	/**
+	 * Filters domains and URLs for resource preloads.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @param array  $urls {
+	 *     Array of resources and their attributes, or URLs to print for resource preloads.
+	 *
+	 *     @type array ...$0 {
+	 *         Array of resource attributes.
+	 *
+	 *         @type string $href        URL to include in resource preloads. Required.
+	 *         @type string $as          How the browser should treat the resource
+	 *                                   (`script`, `style`, `image`, `document`, etc).
+	 *         @type string $crossorigin Indicates the CORS policy of the specified resource.
+	 *         @type string $type        Type of the resource (`text/html`, `text/css`, etc).
+	 *         @type string $media       Accepts media types or media queries. Allows responsive preloading.
+	 *         @type string $imagesizes  Responsive source size to the source Set.
+	 *         @type string $imagesrcset Responsive image sources to the source set.
+	 *     }
+	 * }
+	 */
+	$urls = apply_filters( 'wp_preload_links', array() );
+
+	if ( ! is_array( $urls ) ) {
+		return;
+	}
+
+	$unique_urls = array();
+	foreach ( $urls as $url ) {
+		if ( ! is_array( $url ) ) {
+			continue;
+		}
+
+		$atts = $url;
+		if ( isset( $url['href'] ) ) {
+			$url          = $url['href'];
+			$atts['href'] = $url;
+			if ( isset( $unique_urls[ $url ] ) ) {
+				continue;
+			}
+			$unique_urls[ $url ] = $atts;
+			// Media can use imagesrcset and not href.
+		} elseif ( ! isset( $url['href'] )
+				&& ( 'image' === $url['as'] )
+				&& ( isset( $url['imagesrcset'] ) || isset( $url['imagesizes'] ) )
+		) {
+			if ( isset( $unique_urls[ $url['imagesrcset'] ] ) ) {
+				continue;
+			}
+			$unique_urls[ $url['imagesrcset'] ] = $atts;
+		} else {
+			continue;
+		}
+
+		foreach ( $unique_urls as $atts ) {
+			$html = '';
+
+			foreach ( $atts as $attr => $value ) {
+				if ( ! is_scalar( $value )
+					|| (
+							! in_array( $attr, array( 'as', 'crossorigin', 'href', 'imagesrcset', 'imagesizes', 'type', 'media' ), true )
+							&& ! is_numeric( $attr )
+					)
+				) {
+
+					continue;
+				}
+
+				// imagesrcset only usable when preloading image, ignore otherwise.
+				if ( ( 'imagesrcset' === $attr ) && ( ! isset( $atts['as'] ) || ( 'image' !== $atts['as'] ) ) ) {
+					continue;
+				}
+
+				// imagesizes only usable when preloading image and imagesrcset present, ignore otherwise.
+				if ( ( 'imagesizes' === $attr )
+					&& ( ! isset( $atts['as'] ) || ( 'image' !== $atts['as'] ) || ! isset( $atts['imagesrcset'] ) )
+				) {
+					continue;
+				}
+
+				$value = ( 'href' === $attr ) ? esc_url( $value, array( 'http', 'https' ) ) : esc_attr( $value );
+
+				if ( ! is_string( $attr ) ) {
+					$html .= " $value";
```

Comment:
should these values (here and L3517) and `esc_attr` applied?

Reply:
what would be the benefit? IMO for reading and clarity sake, I think is better to escape where is it right now ( previous line )

Reply:
generally speaking escaping should be applied at the point of output or as close to it as possible. The reasoning is these lines of code might become separated in the future. Fine to leave for now, not a blocker.